### PR TITLE
Make C++ toolchain an optional dependency of `java_binary`

### PR DESCRIPTION
**Pull request by fmeum**
This allows a `java_binary` to be compiled for a platform with no registered C++ toolchain targets.

Along the way removes unused parameters and outdated comments related to C++ toolchain resolution and usage in Java rules.

### CODE REVIEW

**Comment by fmeum** [comment]
This regressed while https://github.com/bazelbuild/bazel/pull/18262 was in review due to the flip of `--incompatible_enable_cc_toolchain_resolution` (and never worked with platform mappings). I decided to split it off since it doesn't require the rules_java update.

File: src/main/starlark/builtins_bzl/bazel/java/bazel_java_binary.bzl [hunk 1]
```diff
@@ -41,7 +41,11 @@ def _bazel_base_binary_impl(ctx, is_test_rule_class):
     executable = _get_executable(ctx)
 
     feature_config = helper.get_feature_config(ctx)
-    strip_as_default = helper.should_strip_as_default(ctx, feature_config)
+    if feature_config:
+        strip_as_default = helper.should_strip_as_default(ctx, feature_config)
```
Comment (fmeum) [inline src/main/starlark/builtins_bzl/bazel/java/bazel_java_binary.bzl:45]: It seems possible that this branch always evaluates to `False` in Bazel, in which case most of this logic could probably go away.

**Comment by fmeum** [comment]
@meteorcloudy CI is failing because the Windows arm64 JDK can't be fetched in an integration test. What do I need to do to make it available?

**Comment by meteorcloudy** [comment]
Probably here: https://github.com/bazelbuild/bazel/blob/master/src/BUILD#L589

**Comment by fmeum** [comment]
@hvadehra Tests are passing now thanks to @meteorcloudy's advice, this PR is ready for review.

File: src/test/shell/bazel/bazel_java_test_defaults.sh [hunk 2]
```diff
@@ -240,4 +240,65 @@ EOF
   expect_not_log "singlejar/singlejar_local"
 }
 
+function test_executable_java_binary_compiles_for_platform_without_cc_toolchain() {
+  mkdir -p pkg
+  cat > pkg/BUILD.bazel <<'EOF'
+# Choose a platform with a registered JDK for which no registered C++ toolchain
```
Comment (hvadehra) [inline src/test/shell/bazel/bazel_java_test_defaults.sh:246]: Is there something about (windows, arm64, c++) that ensures this assumption will always hold? If not, this test may start passing incorrectly at some point in the future. Could we maybe check that building a cc target fails on this platform?

Comment (fmeum) [reply src/test/shell/bazel/bazel_java_test_defaults.sh]: Good point. I added the check and also construct a fake platform that should make this test more stable as we add support for more platforms.

**Comment by fmeum** [comment]
I had to allow networking in the test as otherwise Bzlmod won't work.

**Review by hvadehra — APPROVED** [review]

**Review by hvadehra — APPROVED** [review]

**Comment by keertk** [comment]
@bazel-io fork 7.0.0

**Comment by iancha1992** [comment]
The changes in this PR have been included in [Bazel 7.0.0 RC2](https://releases.bazel.build/7.0.0/rc2/index.html). Please test out the release candidate and report any issues as soon as possible. If you're using Bazelisk, you can point to the latest RC by setting USE_BAZEL_VERSION=last_rc.
Thanks!

**Pull request closed without merging**